File: 2018/04.py
import math, copy, re, hashlib
import itertools as it

template = '[YYYY-MM-DD HH:MM]'

def parse(row):
	guard = dict()
	ts = row[0:len(template)]
	action = row[len(template) + 1:]

	ts = ts[1:len(ts) - 1]
	[date, time] = ts.split(' ')
	[y, m, d] = list(map(int, date.split('-')))
	[h, min] = list(map(int, time.split(':')))
	t = 100
	ts = min + t * h + t ** 2 * d + t ** 4 * m + t ** 8 * y

	return (ts, action)

def sort_data(data):
	data = [parse(i) for i in data]
	data.sort(key=lambda x: x[0])
	return data

def get_minutes(i, guards):
	i += 1
	res = 0
	t1 = 0
	while i < len(guards) and guards[i][1][0] != 'G':
		ts, act = guards[i]
		if act[0] == 'f':
			t1 = ts % 100
		elif act[0] == 'w':
			res += (ts % 100) - t1 
		i += 1
	# print("HERE1",i, guards[i])
	return (i, res)

def get_id(guard):
	_, action = guard
	r = "#\d+"
	# print(action)
	return re.findall(r, action)[0][1:]

def part_1(data):

	guards = sort_data(data)
	i = 0
	maxi = 0
	g_id = None
	gs = dict()

	while i < len(guards):
		# print(i)
		i_next, minutes = get_minutes(i, guards)
		# print(i)
		if minutes > maxi:
			maxi = minutes
			# print(">>",i, guards[i])
			g_id = get_id(guards[i])
		i = i_next
		
	print(maxi * int(g_id))
	print('END OF PART1')
	return

def part_2(data):
	print('END OF PART2')
	return 


if __name__ == '__main__':
	with open('04_input') as f:
		data = f.read()
		data = data.split('\n')
		# data = list(map(int, data.split()))


	part_1(copy.deepcopy(data))
	part_2(copy.deepcopy(data))
	

File: 2018/04.js

var fs = require('fs');
const { ord, chr, debug, disp, int, float, randint, gcd, lcm, crt, modPow, mod } = require("../lib");
const { range, drange, trange, iter, diter, titer } = require("../lib");
const { PriorityQueue, Queue, set, Stack } = require("../lib");
const { min, max, random, abs, ceil, floor, log, log10, log2, round, sign, sin, cos, tan, asin, acos, atan, atan2, sqrt, PI } = Math;
const { isSuperset, or, and, xor, sub } = set;

const sort = (arr, fn = (a, b) => a-b) => {
	arr.sort(fn);
}
const in_ = (a, arr) => arr.indexOf(a) !== -1;

const D4 = [[0,1],[1,0],[0,-1],[-1,0]];
const D8 = [[0,1],[1,1],[1,0],[1,-1],[0,-1],[-1,-1],[-1,0],[-1,1]];
const MOD = 1e9+7;

const string = '[1518-10-31 00:58]';

const parse = (data) => {
	return data.map(r => {
		const time = r.slice(1, string.length - 1);
		const action = r.slice(string.length).trim();
		const [dt, tm] = time.split(' ');
		const [YY, MM, DD] = dt.split('-').map(d => int(d.trim()));
		const [hh, mm] = tm.split(':').map(t => int(t.trim()));
		return { action, YY, MM, DD, hh, mm }
	}).sort((a, b) => a.YY - b.YY || a.MM - b.MM || a.DD - b.DD || a.hh - b.hh || a.mm - b.mm);
}

function part1(data) {

	data = parse(data);

	let id, start, end;
	const guards = {};

	iter(data)(r => {
		const { action, YY, MM, DD, hh, mm } = r;
		if (action.startsWith('Guard')) {
			id = action.split(' ')[1].slice(1);
			if (!guards[id]) {
				guards[id] = new Array(60).fill(0);
			}
		} else if (action.startsWith('falls')) {
			start = mm;
		} else if (action.startsWith('wakes')) {
			if (mm < start) {
				mm += 60;
			}
			range(start, mm)(i => {
				guards[id][i]++;
			});
		}
	});

	let mxg = -1, mxmin = -1;

	Object.keys(guards).forEach(id => {
		const sm = guards[id].reduce((acc, curr) => acc + curr, 0);
		if (sm > mxmin) {
			mxmin = sm;
			mxg = id;
		}
	});

	mxmin = -1;
	let mxi = -1;
	guards[mxg].forEach((min, i) => {
		if (min > mxmin) {
			mxmin = min;
			mxi = i
		}
	});

	debug(int(mxg) * mxi)

	console.log("END OF PART1\n");
	part2(guards);
	return;
}

function part2(guards) {

	let mxg = -1, mxmin = -1;

	Object.keys(guards).forEach(id => {
		let mx = max(...guards[id]);
		if (mx > mxmin) {
			mxmin = mx;
			mxg = id;
		}
	});

	mxmin = -1;
	let mxi = -1;
	guards[mxg].forEach((min, i) => {
		if (min > mxmin) {
			mxmin = min;
			mxi = i;
		}
	})

	debug(int(mxg) * mxi)

	console.log("END OF PART2\n");
	return;
}

function main() {
	let data = fs.readFileSync("04_input").toString("utf-8");
	data = data.split('\n');
	// data = data.split('\n').map(a => Number(a));

	if (Array.isArray(data)) {
		part1(Array.from(data));
		// part2(Array.from(data));
	} else {
		part1(data);
		// part2(data);
	}
}

main();
